### DISCUSSION TITLE
Substructure search with many SMARTS patterns

### DISCUSSION THREAD

**Discussion by f3rmion** [rdkit/rdkit#7277] [Q&A]
Hello community!

I'm currently trying to build a fixed library of SMARTS patterns that I'd like to use to match via substructure search on molecules. I came across the `rdkit.Chem.rdSubstrucLibrary`, but I assume that this is exactly the inverse case with one pattern and many molecules in a library.

Currently, I have a brute force ansatz iterating over the SMARTS library,

```python
smiles = "C1N=CC=C(C(=O)Cl)C=1"
m = Chem.MolFromSmiles(smiles)

matches = []
for k, v in smarts_library.items():
    smarts_pattern = Chem.MolFromSmarts(v)
    match = m.GetSubstructMatch(smarts_pattern)
    if match:
        matches.append((k, v))
# matches in SMARTS library:
# - AcidHalide_Chloride_Het6 pattern
# - Heterocycle6 pattern
```

which I assume is not very efficient. Has someone experience with screening hundreds of SMARTS pattern on molecules in an efficient way? The aim of this is to create a substructure fingerprint that incorporates fixed identifiers of matched substructures - identifiers in the SMARTS library will not change in the future.

**Comment by bp-kelley** [comment]
You’ll want to use the FilterCatalog

https://www.rdkit.org/docs/source/rdkit.Chem.rdfiltercatalog.html

I’m traveling today, but can get a proper example tomorrow, however the test suite may help you out:

https://github.com/rdkit/rdkit/blob/f5431728db153328d3483db7e0cfa2ab7bdbeb32/Code/GraphMol/FilterCatalog/Wrap/rough_test.py#L62

**Comment by f3rmion** [comment]
Thanks @bp-kelley,

After looking into `FilterCatalog`, I have now a first solutions that creates a catalog

```python
from rdkit import Chem
from rdkit import FilterCatalog

catalog = FilterCatalog.FilterCatalog()
for k, v in smarts_library.items():
    # create RDKit molecule
    m = Chem.MolFromSmarts(v)

    # create SMARTS matcher
    sm = FilterCatalog.SmartsMatcher(m)

    # add entry to FilterCatalog
    entry = FilterCatalog.FilterCatalogEntry(k, sm)
    catalog.AddEntry(entry)

smiles = "C1N=CC=C(C(=O)Cl)C=1"
matches = catalog.GetMatches(Chem.MolFromSmiles(smiles))

for match in matches:
    print(match.GetDescription())
# matches in SMARTS catalog:
# - AcidHalide_Chloride_Het6 pattern
# - Heterocycle6 pattern
```

Is this the solution that you thought about?

Cheers,
Eike
